Implement Enchant User Interface

Asheraf opened the pull request:
<!-- Before you continue, please change "base: stable" to "base: master" and
     enable the setting "[√] Allow edits from maintainers." when creating your
     pull request if you have not already enabled it. -->

<!-- Note: Lines with this <!-- syntax are comments and will not be visible in
     your pull request. You can safely ignore or remove them. -->

### Pull Request Prelude

<!-- Thank you for working on improving Hercules! -->
<!-- Please complete these steps and check the following boxes by putting an `x`
     inside the [brackets] before filing your Pull Request. -->

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

<!-- You can safely ignore the links below:  -->

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

4144 commented on src/map/enchantui.c:
```diff
@@ -0,0 +1,929 @@
+
+static bool enchantui_read_db_libconfig_sub(const struct config_setting_t *it, int n, const char *source)
+{
+	nullpo_retr(false, it);
```
missing nullpo check for source

4144 commented on src/map/packets_struct.h:
```diff
@@ -5751,6 +5751,60 @@ struct PACKET_ZC_ITEM_REFORM_ACK {
+DEFINE_PACKET_HEADER(CZ_REQUEST_RESET_ENCHANT, 0x0b9e);
+#endif // PACKETVER_MAIN_NUM >= 20201118 || PACKETVER_RE_NUM >= 20211103
+
+#if PACKETVER_MAIN_NUM >= 20201118 || PACKETVER_RE_NUM >= 20211103
```
this packet starting from main 2021-02-03, re 2021-11-03

MishimaHaruna commented on src/map/enchantui.c:
```diff
@@ -0,0 +1,930 @@
+	struct config_setting_t *enchant_db = NULL;
+	if ((enchant_db = libconfig->setting_get_member(enchant_conf.root, "enchant_db")) == NULL) {
+		ShowError("%s: can't read %s\n", __func__, filepath);
+		return;
```
This return doesn't go through `libconfig->destroy()`

MishimaHaruna commented on src/map/enchantui.c:
```diff
@@ -0,0 +1,930 @@
+		return false;
+	}
+
+	libconfig->setting_lookup_int(it, "MinRefine", (int *)&ei.MinRefine);
```
`MinRefine` is an 8 bit integer, passing it as an int pointer to setting_lookup_int() may cause it to overwrite memory around it.

It would be better to use a separate int32 variable and then assign to MinRefine.

MishimaHaruna commented on src/map/enchantui.c:
```diff
@@ -0,0 +1,930 @@
+		return false;
+	}
+
+	libconfig->setting_lookup_int(it, "MinGrade", (int *)&ei.MinGrade);
```
Same as above

Asheraf commented:
Addressed the requested changes, the enchant_db files need to be regenerated but I seem to have lost the tool I made to do so.. will try to update them before the next release date 🤞

dastgirp commented on src/map/clif.c:
```diff
@@ -25526,6 +25527,71 @@ static void clif_item_reform_result(struct map_session_data *sd, int index, enum
+static void clif_enchantui_open(struct map_session_data *sd, int64 enchant_group)
+{
+#if PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103
+	clif->open_ui_send2(sd, ZC_ENCHANT_UI, enchant_group);
```
State is not saved?
Does it allow to talk to other NPC/use skill/move while UI is opened?

Asheraf commented:
Rebased this pr, also added the missing state checks and updated the database file. @4144 can you check if the defines for the new zero client are correct.

4144 commented on src/map/clif.c:
```diff
@@ -25573,6 +25574,97 @@ static void clif_item_reform_result(struct map_session_data *sd, int index, enum
+static void clif_enchantui_status(struct map_session_data *sd, enum enchantui_status result, int itemId)
+{
+#if PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+	const int fd = sd->fd;
```
missing nullpo check for sd

4144 commented on src/map/enchantui.c:
```diff
@@ -0,0 +1,935 @@
+ * This file is part of Hercules.
+ * http://herc.ws - http://github.com/HerculesWS/Hercules
+ *
+ * Copyright (C) 2021 Hercules Dev Team
```
Missing 2022 in copyright

4144 commented on src/map/enchantui.h:
```diff
@@ -0,0 +1,131 @@
+ * This file is part of Hercules.
+ * http://herc.ws - http://github.com/HerculesWS/Hercules
+ *
+ * Copyright (C) 2021 Hercules Dev Team
```
and here missing in copyright 2022

4144 reviewed (changes requested):
zero versions look like correct
